### DISCUSSION TITLE
keycloak Andriod implementation

### DISCUSSION THREAD

**Discussion by K-a-n-n-a-n-r** [keycloak/keycloak#34976] [Q&A]
I have integrated Keycloak into my web application, and now I'm working on a mobile application that uses the same Keycloak authentication. Currently, I'm using the web UI for login, which then redirects to my mobile screen. However, I'd like to implement this directly in the mobile app using APIs instead.

Could someone help me with the following API endpoints for mobile integration?

User login authentication API
OTP verification for two-factor authentication
Refresh token API
Logout API
Any guidance on implementing these would be greatly appreciated

**Comment by embesozzi** [comment]
If you follow standards and best practices for authentication, you would implement OpenID Connect. This typically involves opening a browser within the mobile app using an OIDC library like  [AppAuth-Android](https://github.com/openid/AppAuth-Android).

However, if you prefer an API-based approach, the current OAuth 2.0 standard offers the  [ROPC](https://datatracker.ietf.org/doc/html/rfc6749) (Resource Owner Credentials) grant, which is focused on password-based authentication. However, this approach is no longer recommended and will be deprecated in OAuth 2.1.

In the "near" future, we expect to see  [OAuth 2.0 for First-Party Applications](https://datatracker.ietf.org/doc/html/draft-parecki-oauth-first-party-apps), an API-based native authentication approach. While we are still in the early stages, I have some working Proof of Concepts (PoCs) based on this approach. However, the extension is not yet production-ready because there are still some points in the draft to be defined and additional use cases to be implemented.
